File: C4/C4Movie.m
// Copyright © 2012 Travis Kirton
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions: The above copyright
// notice and this permission notice shall be included in all copies or
// substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

#import "C4Movie.h"
#import "C4UIMovieControl.h"

@interface C4Movie() {
    void *rateContext, *currentItemContext, *playerItemStatusContext;
}
@property(nonatomic) NSURL *movieURL;
@property(nonatomic, strong) AVPlayerLayer *playerLayer;
@property(nonatomic, strong) AVPlayer *player;
@property(nonatomic, strong) AVPlayerItem *playerItem;
@end

@implementation C4Movie

+ (instancetype)movieNamed:(NSString *)movieName {
    C4Movie *newMovie = [[C4Movie alloc] initWithMovieName:movieName frame:CGRectZero];
    return newMovie;
}

+ (instancetype)movieNamed:(NSString *)movieName inFrame:(CGRect)movieFrame {
    C4Movie *newMovie = [[C4Movie alloc] initWithMovieName:movieName frame:movieFrame];
    return newMovie;
}

+ (instancetype)movieWithURL:(NSString *)url {
    C4Movie *newMovie = [[C4Movie alloc] initWithURL:[NSURL URLWithString:url] frame:CGRectZero];
    return newMovie;
}

+ (instancetype)movieWithURL:(NSString *)url frame:(CGRect)movieFrame {
    C4Movie *newMovie = [[C4Movie alloc] initWithURL:[NSURL URLWithString:url] frame:movieFrame];
    return newMovie;
}

-(id)initWithMovieName:(NSString *)movieName {
    movieName = [movieName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    self = [self initWithMovieName:movieName frame:CGRectZero];
    return self;
}

-(id)initWithMovieName:(NSString *)movieName frame:(CGRect)movieFrame {
    NSArray *movieNameComponents = [movieName componentsSeparatedByString:@"."];
    self.movieURL = [[NSURL alloc] initFileURLWithPath:[[NSBundle mainBundle] pathForResource:movieNameComponents[0] ofType:movieNameComponents[1]]];
    self = [self initWithURL:self.movieURL frame:movieFrame];
    return self;
}

-(id)initWithURL:(NSURL *)url {
    self = [self initWithURL:url frame:CGRectZero];
    return self;
}

-(id)initWithURL:(NSURL *)url frame:(CGRect)frame {
    self = [super initWithView:[[C4UIMovieControl alloc] initWithFrame:frame]];
    if(self != nil) {
        _volume = 1.0f;
        self.shouldAutoplay = NO;
        _movieURL = url;
        if([_movieURL scheme]) {
            AVURLAsset *asset = [AVURLAsset URLAssetWithURL:_movieURL options:nil];
            C4Assert(asset != nil, @"The asset (%@) you tried to create couldn't be initialized", _movieURL);
            
            NSArray *requestedKeys = @[@"duration", @"playable"];
            for (NSString *key in requestedKeys) {
                NSError *error = nil;
                AVKeyValueStatus keyStatus = [asset statusOfValueForKey:key error:&error];
                if (keyStatus == AVKeyValueStatusFailed) {
                    [self assetFailedToPrepareForPlayback:error];
                    return nil;
                }
            }
            
            if (asset.playable == NO) {
                NSError *error = [NSError errorWithDomain:@"C4Movie asset cannot be played" code:0 userInfo:nil];
                [self assetFailedToPrepareForPlayback:error];
                return nil;
            }
            
            [asset loadValuesAsynchronouslyForKeys:requestedKeys completionHandler: ^(void) {
                dispatch_async( dispatch_get_main_queue(), ^(void) {
                    [self prepareToPlayAsset:asset];
                });
            }];
            [self completeSetupWithAsset:asset frame:frame];
        }
    }
    return self;
}

-(void)completeSetupWithAsset:(AVURLAsset *)asset frame:(CGRect)movieFrame {
    NSArray *assetTracks = [asset tracksWithMediaType:AVMediaTypeVideo];
    
    if(CGRectEqualToRect(movieFrame, CGRectZero)) {
        AVAssetTrack *videoTrack = assetTracks[0];
        _originalMovieSize = videoTrack.naturalSize;
        _originalMovieRatio = _originalMovieSize.width / _originalMovieSize.height;
        movieFrame.size = _originalMovieSize;
        self.frame = movieFrame;
    }
    self.backgroundColor = [UIColor clearColor];
    _constrainsProportions = YES;
    self.player.actionAtItemEnd = AVPlayerActionAtItemEndPause; // currently C4Movie doesn't handle queues
    self.rate = 1.0f;
    
    [self setup];
}

- (CMTime)playerItemDuration {
    AVPlayerItem *thePlayerItem = [_player currentItem];
    if (thePlayerItem.status == AVPlayerItemStatusReadyToPlay) {
        return(_playerItem.duration);
    }
    return(kCMTimeInvalid);
}

- (BOOL)isPlaying {
    return (self.player.rate != 0.0f);
}

-(CGFloat)rate {
    return self.player.rate;
}

-(void)setRate:(CGFloat)rate {
    self.player.rate = rate;
}

- (AVPlayerLayer *)playerLayer {
    return (AVPlayerLayer *)self.view.layer;
}

-(void)assetFailedToPrepareForPlayback:(NSError *)error {
    C4Log(@"The movie you tried to load failed...\n\n%@\n\nConfirm the following:\n1)the URL you used is correct.\n2)make sure your device is connected to the internet",error);
}

-(void)prepareToPlayAsset:(AVURLAsset *)asset {
    rateContext = &rateContext;
    currentItemContext = &currentItemContext;
    playerItemStatusContext = &playerItemStatusContext;
    
    if (self.playerItem != nil) {
        [self.playerItem removeObserver:self forKeyPath:@"status"];
        [[NSNotificationCenter defaultCenter] removeObserver:self
                                                        name:AVPlayerItemDidPlayToEndTimeNotification
                                                      object:self.playerItem];
    }
    
    self.playerItem = [AVPlayerItem playerItemWithAsset:asset];
    
    [self.playerItem addObserver:self
                      forKeyPath:@"status"
                         options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
                         context:playerItemStatusContext];
    
    [self listenFor:AVPlayerItemTimeJumpedNotification andRun:^(NSNotification *n) {
        [self currentTimeChanged];
    }];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(playerItemDidReachEnd:)
                                                 name:AVPlayerItemDidPlayToEndTimeNotification
                                               object:self.playerItem];
    
    if (self.player == nil) {
        [self setPlayer:[AVPlayer playerWithPlayerItem:self.playerItem]];
        _player.actionAtItemEnd = AVPlayerActionAtItemEndNone;
        [self.player addObserver:self
                      forKeyPath:@"currentItem"
                         options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew
                         context:currentItemContext];
    }
    
    NSArray *audioTracks = [asset tracksWithMediaType:AVMediaTypeAudio];
    NSMutableArray *allAudioTrackInputParameters = [@[] mutableCopy];
    for (AVAssetTrack *track in audioTracks) {
        AVMutableAudioMixInputParameters *trackInputParameters =[AVMutableAudioMixInputParameters audioMixInputParameters];
        [trackInputParameters setTrackID:[track trackID]];
        [allAudioTrackInputParameters addObject:trackInputParameters];
    }
    
    AVMutableAudioMix *newAudioMix = [AVMutableAudioMix audioMix];
    [newAudioMix setInputParameters:allAudioTrackInputParameters];
    _audioMix = newAudioMix;
    self.playerItem.audioMix = self.audioMix;
    
    if (self.player.currentItem != self.playerItem) {
        [[self player] replaceCurrentItemWithPlayerItem:self.playerItem];
    }
    //explicitly set the volume here, it needs to be set after the audio mix has been created
    self.volume = self.volume;
}

- (void)observeValueForKeyPath:(NSString*) path
                      ofObject:(id)object
                        change:(NSDictionary*)change
                       context:(void*)context
{
    /* AVPlayerItem "status" property value observer. */
    if (context == playerItemStatusContext)
    {
        
        AVPlayerStatus status = [change[NSKeyValueChangeNewKey] integerValue];
        switch (status)
        {
                /* Indicates that the status of the player is not yet known because
                 it has not tried to load new media resources for playback */
            case AVPlayerStatusUnknown:
                //C4Log(@"AVPlayerStatusUnknown");
                break;
            case AVPlayerStatusReadyToPlay:
                /* Once the AVPlayerItem becomes ready to play, i.e.
                 [playerItem status] == AVPlayerItemStatusReadyToPlay,
                 its duration can be fetched from the item. */
                //C4Log(@"AVPlayerStatusReadyToPlay");
                self.playerLayer.hidden = NO;
                
                /* Set the C4PlayerLayer on the view to allow the AVPlayer object to display
                 its content. */
                [self.playerLayer setPlayer:_player];
                if(self.shouldAutoplay == YES) {
                    [self play];
                }
                [self postNotification:@"movieIsReadyForPlayback"];
                break;
            case AVPlayerStatusFailed:{
                AVPlayerItem *thePlayerItem = (AVPlayerItem *)object;
                [self assetFailedToPrepareForPlayback:thePlayerItem.error];
                C4Assert(NO,@"The asset you tried to load couldn't be prepared for playback, check its filename and type, maybe the compression is wrong...");
            }
                break;
            default:
                C4Assert(NO,@"Strange... the player's status is none of: AVPlayerStatusUnknown, AVPlayerStatusReadyToPlay, AVPlayerStatusFailed");
                break;
        }
    }
    /* AVPlayer "currentItem" property observer.
     Called when the AVPlayer replaceCurrentItemWithPlayerItem:
     replacement will/did occur. */
    else if (context == currentItemContext)
    {
        AVPlayerItem *newPlayerItem = change[NSKeyValueChangeNewKey];
        //C4Log(@"currentItemContext");
        
        /* New player item null? */
        if (newPlayerItem == (id)[NSNull null]) {
        }
        else {
            /* Set the AVPlayer for which the player layer displays visual output. */
            [self.playerLayer setPlayer:self.player];
            
            /* Specifies that the player should preserve the video’s aspect ratio and
             fit the video within the layer’s bounds. */
            self.playerLayer.videoGravity = AVLayerVideoGravityResize;
        }
    }
    else {
        [super observeValueForKeyPath:path ofObject:object change:change context:context];
    }
    
    return;
}

-(void)playerItemDidReachEnd:(NSNotification*)aNotification {
    aNotification = aNotification;
    [self postNotification:@"reachedEnd"];
    if(self.loops) {
        if(self.player.rate < 0.0f)
            [self seekToTime:self.duration];
        else
            [self seekToTime:0.0f];
        [self play];
    }
    [self reachedEnd];
}

-(void)play {
    [self.player play];
}

-(void)pause {
    [self.player pause];
}

-(CGFloat)width {
    return self.bounds.size.width;
}

-(void)setWidth:(CGFloat)width {
    CGRect newFrame = self.frame;
    newFrame.size.width = width;
    if(_constrainsProportions) newFrame.size.height = width/self.originalMovieRatio;
    self.frame = newFrame;
}

-(CGFloat)height {
    return self.frame.size.height;
}

-(void)setHeight:(CGFloat)height {
    CGRect newFrame = self.frame;
    newFrame.size.height = height;
    if(_constrainsProportions) newFrame.size.width = height * self.originalMovieRatio;
    self.frame = newFrame;
}

-(CGSize)size {
    return self.frame.size;
}

-(void)setSize:(CGSize)size {
    CGRect newFrame = CGRectZero;
    newFrame.origin = self.origin;
    newFrame.size = size;
    self.frame = newFrame;
}

-(CGFloat)currentTime {
    return (CGFloat)CMTimeGetSeconds(self.player.currentTime);
}

-(CGFloat)duration {
    return (CGFloat)CMTimeGetSeconds(self.playerItem.duration);
}

/* clamps to the nearest frame, based on the movie's time scale */
-(void)seekToTime:(CGFloat)time {
    CMTime current = self.player.currentTime;
    CMTime newTime = CMTimeMakeWithSeconds(time, current.timescale);
    CMTimeRange timeRange = CMTimeRangeMake(CMTimeMakeWithSeconds(0, current.timescale), self.playerItem.duration);
    CMTimeClampToRange(newTime, timeRange);
    [self.player seekToTime:newTime toleranceBefore:CMTimeMake(1, current.timescale) toleranceAfter:CMTimeMake(1, current.timescale)];
}

-(void)seekByAddingTime:(CGFloat)time {
    [self seekToTime:time + self.currentTime];
}

-(void)reachedEnd {
}

-(void)currentTimeChanged {
}

-(void)setVolume:(CGFloat)volume {
    _volume = volume;
    for(AVMutableAudioMixInputParameters *avmam in self.audioMix.inputParameters) {
        [avmam setVolume:volume atTime:kCMTimeZero];
    }
    self.player.currentItem.audioMix = self.audioMix;
}


#pragma mark Templates

+ (C4Template *)defaultTemplate {
    static C4Template* template;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        template = [C4Template templateFromBaseTemplate:[super defaultTemplate] forClass:self];
    });
    return template;
}

@end

File: C4/UI/C4Movie.swift
// Copyright © 2014 C4
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions: The above copyright
// notice and this permission notice shall be included in all copies or
// substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

import UIKit
import AVFoundation

///This document describes the C4Movie class. You use a C4Movie object to implement the playback of video files, it encapulates an AVQueuePlayer object which handles the loading and control of assets.
///
///The C4Movie class is meant to simplify the addition of videos to your application. It is also a subclass of C4View, and so has all the common animation, interaction and notification capabilities.
///
///A C4Movie’s resizing behaviour is to map itself to the edges of its visible frame. This functionality implicitly uses AVLayerVideoGravityResize as its layer’s default gravity. You can change the frame of the movie from an arbitrary shape back to its original proportion by using its originalSize, originalRatio, or by independently setting either its width or height properties.
public class C4Movie: C4View {
    var player : AVQueuePlayer?
    var currentItem : AVPlayerItem?
    var reachedEndAction : (()->())?
    
    /// Assigning a value of true to this property will cause the receiver to loop at the end of playback.
    ///
    /// The default value of this property is `true`.
    public var loops : Bool = true
    
    /// A variable that provides access to the width of the receiver. Animatable.
    /// The default value of this property is defined by the movie being created.
    /// Assigning a value to this property causes the receiver to change the width of its frame. If the receiver's
    /// `contrainsProportions` variable is set to `true` the receiver's height will change to match the new width.
    public override var width : Double {
        get {
            return Double(view.frame.size.width)
        } set(val) {
            var newSize = C4Size(val,height)
            if constrainsProportions {
                newSize.height = val * height / width
            }
            var rect = self.frame
            rect.size = newSize
            self.frame = rect
        }
    }
    
    /// A variable that provides access to the height of the receiver. Animatable.
    /// The default value of this property is defined by the movie being created.
    /// Assigning a value to this property causes the receiver to change the height of its frame. If the receiver's
    /// `contrainsProportions` variable is set to `true` the receiver's width will change to match the new height.
    public override var height : Double {
        get {
            return Double(view.frame.size.height)
        } set(val) {
            var newSize = C4Size(Double(view.frame.size.width),val)
            if constrainsProportions {
                let ratio = Double(self.size.width / self.size.height)
                newSize.width = val * ratio
            }
            var rect = self.frame
            rect.size = newSize
            self.frame = rect
        }
    }
    
    /// Assigning a value of true to this property will cause the receiver to scale its entire frame whenever its `width` or
    /// `height` variables are set.
    /// The default value of this property is `true`.
    public var constrainsProportions : Bool = true
    
    /// The original size of the receiver when it was initialized.
    public internal(set) var originalSize : C4Size = C4Size(1,1)
    
    /// The original width/height ratio of the receiver when it was initialized.
    public var originalRatio : Double {
        get {
            return originalSize.width / originalSize.height
        }
    }
    
    var movieLayer: AVPlayerLayer {
        get {
            return self.movieView.movieLayer
        }
    }
    
    var movieView: MovieView {
        return self.view as! MovieView
    }
    
    class MovieView : UIView {
        
        var movieLayer: AVPlayerLayer {
            get {
                return self.layer as! AVPlayerLayer
            }
        }
        
        override class func layerClass() -> AnyClass {
            return AVPlayerLayer.self
        }
    }
    
    /// Initializes a new C4Movie using the specified filename from the bundle (i.e. your project).
    ///
    /// - parameter name:	The name of the movie file included in your project.
    public convenience init?(_ filename: String) {
        guard let url = NSBundle.mainBundle().URLForResource(filename, withExtension: nil) else {
            return nil
        }
        
        let asset = AVAsset(URL: url)
        let tracks = asset.tracksWithMediaType(AVMediaTypeVideo)
        
        //grab the movie track and size
        let movieTrack = tracks[0]
        let size = C4Size(movieTrack.naturalSize)
        
        self.init(frame: C4Rect(0,0,Double(size.width),Double(size.height)))
        //initialize player with item
        let newPlayer = AVQueuePlayer(playerItem: AVPlayerItem(asset: asset))
        newPlayer.actionAtItemEnd = .Pause
        currentItem = newPlayer.currentItem
        player = newPlayer
        
        //runs an overrideable method for handling the end of the movie
        on(event: AVPlayerItemDidPlayToEndTimeNotification) {
            self.handleReachedEnd()
        }
        
        //movie view's player
        movieLayer.player = player
        movieLayer.videoGravity = AVLayerVideoGravityResize
        
        originalSize = self.size
    }
    
    /// Initializes a new C4Movie using the specified frame.
    ///
    /// - parameter frame:	The frame of the new movie object.
    public init(frame: C4Rect) {
        super.init()
        self.view = MovieView(frame: CGRect(frame))
    }
    
    
    /// Begins playback of the current item.
    ///
    /// This is the same as setting rate to 1.0.
    public func play() {
        guard let p = player else {
            print("The current movie's player is not properly initialized")
            return
        }
        p.play()
    }
    
    /// Pauses playback.
    ///
    /// This is the same as setting rate to 0.0.
    public func pause() {
        guard let p = player else {
            print("The current movie's player is not properly initialized")
            return
        }
        p.pause()
    }
    
    /// Stops playback.
    ///
    /// This is the same as setting rate to 0.0 and resetting the current time to 0.
    public func stop() {
        guard let p = player else {
            print("The current movie's player is not properly initialized")
            return
        }
        p.seekToTime(CMTimeMake(0,1))
        p.pause()
    }
    
    /// Called at the end of playback (i.e. when the movie reaches its end).
    ///
    /// You can override this function to add your own custom actions.
    ///
    /// Default behaviour: if the movie should loop then the method calls `stop()` and `play()`.
    public func reachedEnd(action: ()->()) {
        reachedEndAction = action
    }
    
    func handleReachedEnd() {
        if self.loops {
            self.stop()
            self.play()
        }
        reachedEndAction?()
    }
}